source/daplink/interface/swd_host_ca.c

```diff
@@ -0,0 +1,949 @@
+}
+
+// Write debug port register
+uint8_t swd_write_dp(uint8_t adr, uint32_t val)
+{
+    uint32_t req;
+    uint8_t data[4];
+    uint8_t ack;
+
+    switch (adr) {
+        case DP_SELECT:
+            if (dap_state.select == val) {
+                return 1;
+            }
+
+            dap_state.select = val;
+            break;
+
+        default:
+            break;
+    }
+
+    req = SWD_REG_DP | SWD_REG_W | SWD_REG_ADR(adr);
+    int2array(data, val, 4);
+    ack = swd_transfer_retry(req, (uint32_t *)data);
+    return (ack == 0x01);
+}
+
+// Read access port register.
+uint8_t swd_read_ap(uint32_t adr, uint32_t *val)
+{
+    uint8_t tmp_in, ack;
+    uint8_t tmp_out[4];
+    uint32_t tmp;
+    uint32_t apsel = adr & 0xff000000;
+    uint32_t bank_sel = adr & APBANKSEL;
+
+    if (!swd_write_dp(DP_SELECT, apsel | bank_sel)) {
+        return 0;
+    }
+
+    tmp_in = SWD_REG_AP | SWD_REG_R | SWD_REG_ADR(adr);
+    // first dummy read
+    swd_transfer_retry(tmp_in, (uint32_t *)tmp_out);
+    ack = swd_transfer_retry(tmp_in, (uint32_t *)tmp_out);
+    *val = 0;
+    tmp = tmp_out[3];
+    *val |= (tmp << 24);
+    tmp = tmp_out[2];
+    *val |= (tmp << 16);
+    tmp = tmp_out[1];
+    *val |= (tmp << 8);
+    tmp = tmp_out[0];
+    *val |= (tmp << 0);
+    return (ack == 0x01);
+}
+
+// Write access port register
+uint8_t swd_write_ap(uint32_t adr, uint32_t val)
+{
+    uint8_t data[4];
+    uint8_t req, ack;
+    uint32_t apsel = adr & 0xff000000;
+    uint32_t bank_sel = adr & APBANKSEL;
+
+    if (!swd_write_dp(DP_SELECT, apsel | bank_sel)) {
+        return 0;
+    }
+
+    switch (adr) {
+        case AP_CSW:
+            if (dap_state.csw == val) {
+                return 1;
+            }
+
+            dap_state.csw = val;
+            break;
+
+        default:
+            break;
+    }
+
+    req = SWD_REG_AP | SWD_REG_W | SWD_REG_ADR(adr);
+    int2array(data, val, 4);
+
+    if (swd_transfer_retry(req, (uint32_t *)data) != 0x01) {
+        return 0;
+    }
+
+    req = SWD_REG_DP | SWD_REG_R | SWD_REG_ADR(DP_RDBUFF);
+    ack = swd_transfer_retry(req, NULL);
+    return (ack == 0x01);
+}
+
+uint8_t swd_ca_select_state(uint32_t addr) {
+    uint8_t tmp_in[4];
+    uint32_t work_select_state;
+
+    if ((DEBUG_REGSITER_BASE <= addr) && (addr <= DBGCID3)) {
+        work_select_state = SELECT_DBG;
+    } else {
+        work_select_state = SELECT_MEM;
+    }
+    if (select_state != work_select_state) {
+        // SELECT
+        select_state = work_select_state;
+        int2array(tmp_in, select_state, 4);
+        if (swd_transfer_retry(0x08, (uint32_t *)tmp_in) != 0x01) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+
+// Write 32-bit word aligned values to target memory using address auto-increment.
+// size is in bytes.
+static uint8_t swd_write_block(uint32_t address, uint8_t *data, uint32_t size)
+{
+    uint8_t tmp_in[4], req;
+    uint32_t size_in_words;
+    uint32_t i, ack;
```

Comment:
Testing failed because the following warning:
```
compiling swd_host_ca.c...
..\..\..\source\daplink\interface\swd_host_ca.c(265): warning: C4017W: ack may be used before being set
    uint32_t i, ack;
..\..\..\source\daplink\interface\swd_host_ca.c: 1 warning, 0 errors
```